engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoMultiValue.java

```diff
@@ -545,24 +553,26 @@ public final Object findTypeOrNull(
         Type type, EnsoMultiValue mv, boolean reorderOnly, boolean allTypes) {
       var dispatch = mv.dispatch;
       var i = findNode.executeFindIndex(type, dispatch);
-      if (i == -1 && allTypes) {
-        var extraIndex = findNode.executeFindIndex(type, mv.extra);
-        i = extraIndex == -1 ? -1 : dispatch.typesLength() + extraIndex;
+      if (i == -1) {
+        if (allTypes) {
+          var extraIndex = findNode.executeFindIndex(type, mv.extra);
+          if (extraIndex != -1) {
+            if (extraIndex < mv.firstDispatch) {
+              i = extraIndex;
+            } else {
+              var rem = extraIndex - mv.firstDispatch;
+              i = mv.firstDispatch + dispatch.typesLength() + rem;
+              assert i < mv.values.length;
+            }
+          }
+        }
+      } else {
+        i += mv.firstDispatch;
       }
       if (i != -1) {
         if (reorderOnly) {
-          var copyTypes = allTypesWith.executeAllTypes(dispatch, mv.extra);
-          if (i == 0 && dispatch.typesLength() == 1) {
-            return newNode.newValue(copyTypes, 1, mv.values);
-          } else {
-            copyTypes = copyTypes.clone();
-            var copyValues = mv.values.clone();
```

Comment:
The idea behind this PR's speed up is to avoid `mv.values.clone()` when we only need to _select new single pivot_ value. Instead we give each `EnsoMultiValue` a `firstDispatch` index into the array that identifies the location of the currently selected type. Then ths `reorderOnly` operation can **share the same** `mv.values` array with the new `EnsoMultiValue`.